### PR TITLE
perform more aggressive optimization

### LINKED ISSUE #1390: Some specialized class instances are not inlined properly.
As far as I can tell, the specialized methods `==` and `+` are implemented according to the same pattern for complex floats. However only the specialized `==` is inlined as expected.

```lisp
COALTON-USER> (coalton-toplevel
                (define x (lisp (Complex Double-Float) () #C(1.0d0 2.0d0)))
                (define y (lisp (Complex Double-Float) () #C(3.0d0 3.0d0))))
; No value
COALTON-USER> (coalton-codegen-types
                (monomorphize) (define z (+ x y))
                (monomorphize) (define p (== x y)))
(COMMON-LISP:PROGN
 (COMMON-LISP:DECLAIM
  (SB-EXT:MUFFLE-CONDITIONS SB-KERNEL:REDEFINITION-WARNING))
 (COALTON-IMPL/GLOBAL-LEXICAL:DEFINE-GLOBAL-LEXICAL P COMMON-LISP:BOOLEAN)
 (COMMON-LISP:SETF P
                     (COMMON-LISP:LET ((#:B-11129489 Y))
                       (COMMON-LISP:DECLARE
                        (COMMON-LISP:IGNORABLE #:B-11129489))
                       (COMMON-LISP:LET ((#:A-11119488 X))
                         (COMMON-LISP:DECLARE
                          (COMMON-LISP:IGNORABLE #:A-11119488))
                         (COMMON-LISP:LET ((#:A-11529493 #:A-11119488))
                           (COMMON-LISP:DECLARE
                            (COMMON-LISP:IGNORABLE #:A-11529493))
                           (COMMON-LISP:LET ((#:B-11539494 #:B-11129489))
                             (COMMON-LISP:DECLARE
                              (COMMON-LISP:IGNORABLE #:B-11539494))
                             (COMMON-LISP:THE
                              (COMMON-LISP:VALUES COMMON-LISP:BOOLEAN
                                                  COMMON-LISP:&OPTIONAL)
                              (COMMON-LISP:LET ((COALTON-LIBRARY/MATH/COMPLEX::B
                                                 #:B-11539494)
                                                (COALTON-LIBRARY/MATH/COMPLEX::A
                                                 #:A-11529493))
                                (COMMON-LISP:DECLARE
                                 (COMMON-LISP:TYPE
                                  (COMMON-LISP:COMPLEX
                                   COMMON-LISP:DOUBLE-FLOAT)
                                  COALTON-LIBRARY/MATH/COMPLEX::A
                                  COALTON-LIBRARY/MATH/COMPLEX::B))
                                (COMMON-LISP:VALUES
                                 (COMMON-LISP:= COALTON-LIBRARY/MATH/COMPLEX::A
                                                COALTON-LIBRARY/MATH/COMPLEX::B)))))))))
 (COMMON-LISP:SETF (COMMON-LISP:DOCUMENTATION 'P 'COMMON-LISP:VARIABLE)
                     "P :: BOOLEAN")
 (COALTON-IMPL/GLOBAL-LEXICAL:DEFINE-GLOBAL-LEXICAL Z
                                                    (COMMON-LISP:COMPLEX
                                                     COMMON-LISP:DOUBLE-FLOAT))
 (COMMON-LISP:SETF Z
                     (COMMON-LISP:LET ((#:B-11149492 Y))
                       (COMMON-LISP:DECLARE
                        (COMMON-LISP:IGNORABLE #:B-11149492))
                       (COMMON-LISP:LET ((#:A-11139491 X))
                         (COMMON-LISP:DECLARE
                          (COMMON-LISP:IGNORABLE #:A-11139491))
                         (COALTON-LIBRARY/MATH/COMPLEX::COMPLEX-PLUS
                          COALTON-LIBRARY/MATH/COMPLEX::|INSTANCE/COMPLEX DOUBLE-FLOAT|
                          #:A-11139491 #:B-11149492))))
 (COMMON-LISP:SETF (COMMON-LISP:DOCUMENTATION 'Z 'COMMON-LISP:VARIABLE)
                     "Z :: (COMPLEX DOUBLE-FLOAT)")
 (COMMON-LISP:DECLAIM
  (SB-EXT:UNMUFFLE-CONDITIONS SB-KERNEL:REDEFINITION-WARNING))
 (COMMON-LISP:VALUES))
```

### PR DESCRIPTION
**Pull request by stylewarning**
This PR does a variety of things, but the most important are as follows:

- During optimization, we propagate known dictionaries to their use sites.
- During rewriting (inlining or specialization), we track if we do a rewrite, and if we do, we let the optimizer run again.
- During inlining, we ensure the term being replaced and the replacement itself have unified types.
- We make fixnum arithmetic optimized and unsafe.

Fix #1390

### CODE REVIEW

File: library/iterator.lisp [hunk 1]
```diff
@@ -191,6 +191,16 @@ iterator is empty."
     "An iterator which begins at zero and counts up through and including LIMIT."
     (up-to (+ 1 limit)))
 
+
+  ;; All the haranguing below is so that we don't overflow a bounded
+  ;; type in a range-decreasing call. We generally assume a lower
+  ;; bound (e.g., 0 for unsigned types) is more common than an upper
+  ;; bound.
+  (define-type RangeStatus
```
Comment (YarinHeffes) [inline library/iterator.lisp:199]: Optional:
```suggestion
  (repr :enum)
  (define-type RangeStatus
```

Comment (stylewarning) [reply library/iterator.lisp:200]: done

File: library/iterator.lisp [hunk 2]
```diff
@@ -201,16 +212,36 @@ iterator is empty."
...
-        "STEP ~a should be positive and non-zero in RANGE-INCREASING"
+        "STEP ~a should be positive and non-zero in RANGE-DECREASING"
         step)
-    ;; FIXME: avoid underflow in the DONE? test
-    (recursive-iter ((flip -) step)
-                    (fn (n) (>= end (+ n step))) ; like (>= (- end step)), but without potential underflow
-                    (- start step)               ; begin after START
-                    ))
+    (let ((end+step (+ end step)))
+      (if (< start end+step)
```
Comment (YarinHeffes) [inline library/iterator.lisp:221]:
```suggestion
      (if (< start end)
```

File: library/iterator.lisp [hunk 3]
```diff
@@ -201,16 +212,36 @@ iterator is empty."
...
+          empty
+          (let ((start-step (- start step))
+                (next (cell:new start-step))
+                (status (cell:new (if (< start-step end+step)
+                                      RangeLast
+                                      RangeContinue))))
```
Comment (YarinHeffes) [inline library/iterator.lisp:227]:
```suggestion
          (let ((next (cell:new start))
                (status (cell:new (if (< start end+step)
                                      RangeLast
                                      RangeContinue))))
```

File: tests/iterator-tests.lisp [hunk 4]
```diff
@@ -100,6 +100,12 @@
                       (iter:repeat-for "foo" 10)))))
 
 (define-test iter-downfrom ()
+  (is (== (the (List UFix) Nil)
+          (iter:collect! (iter:down-from 0))))
```
Comment (YarinHeffes) [inline tests/iterator-tests.lisp:104]: Maybe we can add 1-3 tests for `iter:range-decreasing` close to the boundary, such as 
```lisp
(iter:range-decreasing (the UFix 2) 5 0)
(iter:range-decreasing (the UFix 2) 3 0)
(iter:range-decreasing (the UFix 2) 1 0)
```

**Review by YarinHeffes — CHANGES_REQUESTED** [review]
Minor changes in logic in `range-decreasing`.

File: library/iterator.lisp (hunk 3 above)
Comment (YarinHeffes) [reply library/iterator.lisp:227]: Maybe I misunderstood the function. I see now that the docstring says "starts below START", in which case your implementation is correct and you can ignore these two comments.

Comment (stylewarning) [reply library/iterator.lisp:227]: yes it should be the same as (reverse (range n)) basically.

File: tests/iterator-tests.lisp (hunk 4 above)
Comment (stylewarning) [reply tests/iterator-tests.lisp:104]: done

**Review by YarinHeffes — APPROVED** [review]
